### PR TITLE
feat(pipelines): Add endpoint to search for pipeline executions.

### PR DESCRIPTION
**Pull request by joonlim**
Also related:
https://github.com/spinnaker/echo/pull/277
https://github.com/spinnaker/gate/pull/560

### CODE REVIEW

File: orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java [hunk 1]
```diff
@@ -439,6 +460,84 @@ public void delete(@Nonnull ExecutionType type, @Nonnull String id) {
...
+                return p.zrevrangeByScore(key, criteria.getBuildTimeEndBoundary(), criteria.getBuildTimeStartBoundary());
+              }
+            );
+
+    /*
+     * Construct an observable that will retrieve pipelines from the primary redis
+     */
+    List<String> currentPipelineIds = filteredPipelineIdsByDelegate.getOrDefault(redisClientDelegate, new ArrayList<>());
+
+    Observable<Execution> currentObservable = retrieveObservable(
```
Comment (lwander) [inline orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java]: This seems to be duplicated with the call to the previous redis -- can we pull this into a helper?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 2]
```diff
@@ -425,6 +575,83 @@ class TaskController {
...
+      }
+      return true
+    } else if (Collection.isInstance(object) && Collection.isInstance(subset)) {
+      // object matches subset if:
+      // - object contains a unique item that matches each item in subset
+      //   * this means that an item in subset may not match to more than one item in object, which
+      //     means that we should check every permutation of object to avoid greedily stopping
+      //     on the first item that matches.
+      //     e.g., Given,
+      //             object: [ { "name": "a", "version": "1" }, { "name" } ]
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: is { "name" } valid json?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 3]
```diff
@@ -425,6 +575,83 @@ class TaskController {
...
+      //           Without checking all permutations, this will match:
+      //             subset: [ { "name": "a", "version": "1" }, { "name" } ]
+      //           but will not match:
+      //             subset: [ { "name": "a" }, { "name", "version": "1" } ]
+      //           because the first item in subset will greedily match the first item in object,
+      //           leaving the second items in both, which do not match. This is fixed by checking
+      //           all permutations of object.
+      Collection<List> permutationsOfObjectAsList = Collections2.permutations((Collection) object)
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: Generating all permutations in memory feels very expensive... Is there a way to generate permutations one by one and exit early?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 4]
```diff
@@ -425,6 +575,83 @@ class TaskController {
...
+    List<String> applicationNames = front50Service.getAllApplications()*.name as List<String>
+    List<String> pipelineConfigIds = applicationNames.stream()
+      .map{ applicationName -> front50Service.getPipelines(applicationName, false)*.id as List<String> }
+      .flatMap{ c -> c.stream() }
+      .collect(Collectors.toList())
+
+    return pipelineConfigIds
+  }
+
+  private static boolean recursivelyCheckIfObjectFieldsMatchesAllSubsetFields(Object object, Object subset) {
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: I think this needs some unit tests on its own.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 5]
```diff
@@ -425,6 +575,83 @@ class TaskController {
...
+      for (Object key : subsetAsMap.keySet()) {
+        if (!recursivelyCheckIfObjectFieldsMatchesAllSubsetFields(objectAsMap.get(key), subsetAsMap.get(key))) {
+          return false
+        }
+      }
+      return true
+    } else if (Collection.isInstance(object) && Collection.isInstance(subset)) {
+      // object matches subset if:
+      // - object contains a unique item that matches each item in subset
+      //   * this means that an item in subset may not match to more than one item in object, which
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: Aren't lists generally considered to be ordered in JSON? Do we really have to check all permutations?

Also, if we allow subsets of maps, why not allow for subsets of lists?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 6]
```diff
@@ -179,6 +189,135 @@ class TaskController {
...
+
+        // Compare each execution's trigger to input triggerParams
+        Map triggerAsMap = mapper.convertValue(it.getTrigger(), Map.class)
+        return recursivelyCheckIfObjectFieldsMatchesAllSubsetFields(triggerAsMap, triggerParams)
+      }
+    }).subscribeOn(Schedulers.io())
+      .toList()
+      .toBlocking()
+      .single()
+      .sort(reverseBuildTime)
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: Just to confirm -- the oldest pipelines will be returned first?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 7]
```diff
@@ -179,6 +189,135 @@ class TaskController {
...
+    }
+
+    List<Execution> rval
+    if (startIndex >= pipelineExecutions.size()) {
+      rval = []
+    } else {
+      rval = pipelineExecutions.subList(startIndex, Math.min(pipelineExecutions.size(), startIndex + size))
+    }
+
+    if (!expand) {
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: Often this kind of parameter exists to make the endpoint more performant -- could you add a TODO to perhaps handle this before making the expensive calls to filter & retrieve all executions?

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 8]
```diff
@@ -179,6 +189,135 @@ class TaskController {
...
+    @RequestParam(value = "eventId", required = false) String eventId,
+    @RequestParam(value = "trigger", required = false) String encodedTriggerParams,
+    @RequestParam(value = "triggerTimeStartBoundary", defaultValue = "0") long triggerTimeStartBoundary,
+    @RequestParam(value = "triggerTimeEndBoundary", defaultValue = "9223372036854775807" /* Long.MAX_VALUE */) long triggerTimeEndBoundary,
+    @RequestParam(value = "statuses", required = false) String statuses,
+    @RequestParam(value = "startIndex", defaultValue =  "0") int startIndex,
+    @RequestParam(value = "size", defaultValue = "10") int size,
+    @RequestParam(value = "reverse", defaultValue = "false") boolean reverse,
+    @RequestParam(value = "expand", defaultValue = "false") boolean expand
+  ) {
```
Comment (jtk54) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: May be worth it to break into helpers for readability.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 9]
```diff
@@ -179,6 +189,135 @@ class TaskController {
...
+   * 0. If this value is missing, it is defaulted to 10.
+   * @param reverse (optional) Reverses the resulting list before it is paginated. If this value is
+   * missing, it is defaulted to false.
+   * @param expand (optional) Expands each execution object in the resulting list. If this value is
+   * missing, it is defaulted to false.
+   * @return
+   */
+  @PreAuthorize("hasPermission(#application, 'APPLICATION', 'READ')")
+  @RequestMapping(value = "/applications/{application}/pipelines/{triggerType}", method = RequestMethod.GET)
+  List<Execution> searchForPipelineExecutions(
```
Comment (jtk54) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: I suggest `searchPipelinesByTrigger` or something similar.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy (hunk 4 above)
Comment (jtk54) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: This function seems complex enough to warrant its own test case in a spec.

Comment (jtk54) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: I'd make this name a bit terser, e.g. `checkObjectMatchesSubset` and describe the recursion algo a bit in a javadoc comment.

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy (hunk 6 above)
Comment (joonlim) [reply orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: By default, newer pipelines will be returned first (sorts by build time in reverse order).

**Comment by joonlim** [comment]
@jtk54 @lwander @ewiseblatt Please review again when you get a chance

File: orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java [hunk 10]
```diff
@@ -96,6 +75,23 @@
     new TypeReference<List<SystemNotification>>() {
     };
 
+  private static String GET_EXECUTIONS_FOR_PIPELINE_CONFIG_IDS_SCRIPT = "" +
+    "local executions = {}\n" +
```
Comment (lwander) [inline orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java]: I think the "prefered" way we've been formatting multi-line strings is like so:

```java
String.join("\n", 
"line one",
"line two",
"line three")
```

File: orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java [hunk 11]
```diff
@@ -729,6 +659,19 @@ public boolean hasEntityTags(@Nonnull ExecutionType type, @NotNull String id) {
     });
   }
 
+  private Map<String, String> buildExecutionMapFromRedisResponse(List<String> entries) {
+    Map<String, String> map = new HashMap<>();
+    String nextKey = null;
+    for (int i = 0; i < entries.size(); i++) {
```
Comment (lwander) [inline orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java]: I'd validate that entries.size is even

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 12]
```diff
@@ -484,6 +499,36 @@ class TaskController {
...
+      throw new IllegalArgumentException(String.format("size must be > 0: size=%s", size))
+    }
+  }
+
+  private Map decodeTriggerParams(String encodedTriggerParams) {
+    Map triggerParams
+    if (encodedTriggerParams != null) {
+      byte[] decoded = Base64.getDecoder().decode(encodedTriggerParams)
+      String str = new String(decoded, Charset.forName("UTF-8"))
+      triggerParams = mapper.readValue(str, Map.class)
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: I would wrap both the base64 decoding & json parsing in a `try` block

File: orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy [hunk 13]
```diff
@@ -586,70 +632,86 @@ class TaskController {
...
-              objectAsList.remove(objectItem) // make sure to not match the same item more than once
-              matchedItem = true
-              break
-            }
-          }
+    }
 
-          if (!matchedItem) {
-            matchedAllItems = false
+    PermutationGenerator subsetPermutations = new PermutationGenerator(subset)
```
Comment (lwander) [inline orca-web/src/main/groovy/com/netflix/spinnaker/orca/controllers/TaskController.groovy]: Nice

**Pull request merged**